### PR TITLE
Solve universe target issues

### LINKED ISSUE #326: 'universe' target should not be an option for normalization
There are two types of "targets" in bigmetadata; universe and denominator. The denominator targets indicate how the variable can be normalized. The universe targets captures the variable used to calculate the median or per capita measurement. We should remove universe targets from the normalization option. We might also just remove the universe target entirely, because is there any known use for it? I'm not sure.

image

### PR DESCRIPTION
**Pull request by antoniocarlon**
- Restored `universe` target
 - Restored `universe` target for `median` and `average` aggregated measurements
 - Added a check to avoid having `median` and `average` aggregated measurements without `universe` target 
 - Fixing other minor issues

Fixes #326

**Note for the Acceptance:** BLS (US), Australia and Canada (NHS and census) had some `median` and `average` aggregated measurements without `universe` target

### CODE REVIEW

File: tasks/util.py [hunk 1]
```diff
@@ -1665,6 +1669,23 @@ def check_null_columns(self):
...
+        session = current_session()
+        result = session.execute("SELECT c.id, c.aggregate, cc.reltype "
+                                 "FROM observatory.obs_table t "
+                                 "INNER JOIN observatory.obs_column_table ct ON t.id = ct.table_id "
+                                 "INNER JOIN observatory.obs_column c ON ct.column_id = c.id "
+                                 "FULL JOIN observatory.obs_column_to_column cc ON c.id = cc.source_id "
+                                 "WHERE t.tablename = '{table}' "
+                                 "AND c.aggregate IN ('average', 'median') "
+                                 "AND (reltype IS NULL OR reltype <> 'universe')".format(
+                                     table=self.output()._tablename)).fetchall()
```
Comment (javitonino) [inline tasks/util.py]: I think this query also returns an error if an average/median column has a denominator. Does this case make sense?

I think it does (dividing an average of a class over the average of the total population, e.g: average income of engineers / average income).

File: tests/test_util.py [hunk 2]
```diff
@@ -139,19 +139,21 @@ def test_column_target_many_inits():
...
 @with_setup(setup, teardown)
 def test_table_target_many_inits():
-    pop_col = ColumnTarget(OBSColumn(
-        id='tests.population',
-        type='Numeric',
-        name="Total Population",
-        description='The total number of all',
-        aggregate='sum',
-        weight=10), FakeTask())
+    total_pop = OBSColumn(
```
Comment (javitonino) [inline tests/test_util.py]: Now that I see test changes, why don't we try to add a test for this?

File: tasks/util.py [hunk 3]
```diff
@@ -1665,6 +1669,23 @@ def check_null_columns(self):
             raise ValueError('The following columns of the table "{table}" contain only NULL values: {columns}'.format(
                 table=self.output().table, columns=', '.join([x[0] for x in result])))
 
+    def check_universe_in_aggregations(self):
```
Comment (javitonino) [inline tasks/util.py]: Now that I see test changes, why don't we try to add a test for this check?

Comment (antoniocarlon) [reply tasks/util.py]: Added test forcing the `ValueError` to raise when there is a `median` or an `average` aggregation without `universe` denominator

File: tasks/util.py (hunk 1 above)
Comment (antoniocarlon) [reply tasks/util.py]: Currently, there isn't any case like this in the ETL but it makes sense, having a median/average with both a universe and a denominator shouldn's trigger the check.

**Comment by antoniocarlon** [comment]
1. This PR should fix all the median and average aggregated measurements without `universe` target. The note for the acceptance tries to help about what to test
2. There's no need for changes in the extension as I have updated the `OBS_Meta` generation (also, see the comments [here](https://github.com/CartoDB/bigmetadata/issues/326))

File: tasks/carto.py [hunk 4]
```diff
@@ -490,6 +500,10 @@ class OBSMeta(Task):
         AND numer_ctag.column_id = numer_c.id
         AND numer_ctag.tag_id = numer_tag.id
         AND numer_c.id = leftjoined_denoms.all_numer_id
+        AND NOT EXISTS (
+          SELECT 1 FROM agg_wo_universe u
+          WHERE u.id = numer_c.id
+        )
```
Comment (antoniocarlon) [inline tasks/carto.py]: I have also tested the possible loss of performance of adding this check and it's insignificant

Comment (javitonino) [reply tasks/carto.py]: Nice to see. Not sure if we want to keep this in this query as it is redundant with the tasks check. My concern is not performance but readability, my head almost exploded reading this. If you prefer to keep it, let's add some comments, at least.

File: tasks/carto.py [hunk 5]
```diff
@@ -490,6 +500,10 @@ class OBSMeta(Task):
         AND numer_ctag.column_id = numer_c.id
         AND numer_ctag.tag_id = numer_tag.id
         AND numer_c.id = leftjoined_denoms.all_numer_id
+        AND NOT EXISTS (
```
Comment (javitonino) [inline tasks/carto.py]: We may want to add a check to remove the non-denominated versions of the universe numerators:

`AND NOT (numer_agg IN ('median', 'average') AND denom_id IS NULL)` or something along those lines.

**Edit**: There are no NULL denom_id entries in the `obs_meta` table (except if a column has no denominators at all). This idea doesn't work.

File: tasks/util.py [hunk 6]
```diff
@@ -1665,6 +1669,24 @@ def check_null_columns(self):
...
+        session = current_session()
+        result = session.execute("SELECT c.id, c.aggregate, STRING_AGG(COALESCE(cc.reltype,''), ',') reltype "
+                                 "FROM observatory.obs_table t "
+                                 "INNER JOIN observatory.obs_column_table ct ON t.id = ct.table_id "
+                                 "INNER JOIN observatory.obs_column c ON ct.column_id = c.id "
+                                 "FULL JOIN observatory.obs_column_to_column cc ON c.id = cc.source_id "
+                                 "WHERE t.tablename = '{table}' "
+                                 "AND c.aggregate IN ('average', 'median') "
+                                 "GROUP BY 1, 2 "
+                                 "HAVING LOWER(STRING_AGG(COALESCE(cc.reltype,''), ',')) NOT LIKE '%universe%'".format(
```
Comment (javitonino) [inline tasks/util.py]: You could make this prettier with `ARRAY_AGG` and `ANY`

**Comment by javitonino** [comment]
Like a charm, merging.

Small issues:
- au.data is very slow generating columns due to the yields. It seems a return+list comprehension is faster
- I manually reduced the version for some columns in order to reimport, since this was missing some obscure dumps (eg: ACS quantiles need to be bumped in order to be able to easily run the normal columns).

Al so, I run into the slow meta generation again, fixed it with an analyze. See https://github.com/CartoDB/bigmetadata/issues/311

**Pull request merged**